returnn/datasets/map.py

```diff
@@ -0,0 +1,210 @@
+"""
+Provides :class:`MapDatasetBase`
+"""
+
+from returnn.datasets.basic import DatasetSeq
+from returnn.datasets.cached2 import CachedDataset2
+from returnn.util.basic import OptionalNotImplementedError
+
+
+class MapDatasetBase(object):
+  """
+  This dataset can be used as template to implement user-side Datasets, where the data can be access in arbitrary order.
+  For global sorting, the length information needs to be known beforehand.
+  """
+
+  def __init__(self, num_outputs=None):
+    """
+    :param dict num_outputs: definition of the data as constructor parameters of a Data object
+    """
+    self.num_outputs = num_outputs or {}
+
+  def __len__(self):
+    """
+    :return: total number of sequences in the dataset
+    :rtype: int
+    """
+    raise NotImplementedError
+
+  def __getitem__(self, seq_idx):
+    """
+    This function does the actual data loading, the order can be arbitrary.
+
+    :param int seq_idx:
+    :return: The content of a single dataset entry
+    :rtype: dict[str,numpy.array]
+    """
+    raise NotImplementedError
+
+  def get_seq_len(self, seq_idx):
+    """
+    This optional function provides the sequence length for the `seq_ordering` parameter.
+    If not specified only a limited set of options is available.
+
+    :param seq_idx:
+    :return:
+    :rtype: int|None
```

Comment:
Is this correct? If implemented, it should always return int, right?
```suggestion
    :rtype: int
```